File: apps/pwa/src/domain/insulin.ts
import type { GlucoseTrend } from "./models";

export interface DoseCalculation {
  correction: number;
  carbohydrate_coverage: number;
  trend_adjustment: number;
  total: number;
  suggested: number;
}

const TREND_ADJUSTMENTS: Readonly<Record<GlucoseTrend, readonly [number, number, number, number]>> = {
  SUBINDO_RAPIDO: [2, 2, 1, 1],
  SUBINDO: [1, 1, 1, 0],
  ESTAVEL: [0, 0, 0, 0],
  CAINDO: [-1, -1, -1, 0],
  CAINDO_RAPIDO: [-2, -2, -1, 0],
  NAO_INFORMADA: [0, 0, 0, 0]
};

export function roundHalfAwayFromZero(value: number): number {
  return Math.sign(value) * Math.floor(Math.abs(value) + 0.5);
}

export function trendAdjustment(trend: GlucoseTrend, correctionFactor: number): number {
  const adjustments = TREND_ADJUSTMENTS[trend];
  if (correctionFactor < 25) {
    return adjustments[0];
  }
  if (correctionFactor < 50) {
    return adjustments[1];
  }
  if (correctionFactor <= 75) {
    return adjustments[2];
  }
  return adjustments[3];
}

export interface DoseInput {
  glucose: number;
  carbohydrates: number;
  target_glucose: number;
  correction_factor: number;
  carbohydrate_ratio: number;
  trend: GlucoseTrend;
}

export function calculateSuggestedDose(input: DoseInput): DoseCalculation {
  const {
    glucose,
    carbohydrates,
    target_glucose: targetGlucose,
    correction_factor: correctionFactor,
    carbohydrate_ratio: carbohydrateRatio,
    trend
  } = input;

  if (glucose <= 0 || carbohydrates < 0) {
    throw new Error("Glicemia e carboidratos inválidos.");
  }
  if (correctionFactor <= 0 || carbohydrateRatio <= 0) {
    throw new Error("Fatores de insulina devem ser maiores que zero.");
  }

  const correction = (glucose - targetGlucose) / correctionFactor;
  const carbohydrateCoverage = carbohydrates / carbohydrateRatio;
  const adjustment = trendAdjustment(trend, correctionFactor);
  const total = correction + carbohydrateCoverage + adjustment;

  return {
    correction,
    carbohydrate_coverage: carbohydrateCoverage,
    trend_adjustment: adjustment,
    total,
    suggested: Math.max(0, roundHalfAwayFromZero(total))
  };
}

File: apps/cli/src/glicia/domain/insulin.py
"""Cálculo determinístico de insulina, sem dependência da IA."""

from __future__ import annotations

import math
from dataclasses import dataclass

from glicia.domain.models import GlucoseTrend

TREND_ADJUSTMENTS: dict[GlucoseTrend, tuple[int, int, int, int]] = {
    # Limites conservadores configurados para o projeto: a seta rápida sobe
    # no máximo 2 U, a seta simples 1 U e as quedas reduzem no máximo 2 U.
    GlucoseTrend.RISING_FAST: (2, 2, 1, 1),
    # Regra conservadora configurada para o projeto: a seta simples nunca
    # acrescenta mais que 1 U, mesmo nas faixas de maior resistência.
    GlucoseTrend.RISING: (1, 1, 1, 0),
    GlucoseTrend.STABLE: (0, 0, 0, 0),
    GlucoseTrend.FALLING: (-1, -1, -1, 0),
    GlucoseTrend.FALLING_FAST: (-2, -2, -1, 0),
    GlucoseTrend.NOT_INFORMED: (0, 0, 0, 0),
}


@dataclass(frozen=True)
class DoseCalculation:
    correction: float
    carbohydrate_coverage: float
    trend_adjustment: int
    total: float
    suggested: int


def round_half_away_from_zero(value: float) -> int:
    return int(math.copysign(math.floor(abs(value) + 0.5), value))


def trend_adjustment(trend: GlucoseTrend, correction_factor: float) -> int:
    """Ajuste proposto para FreeStyle Libre, por faixa de FC em mg/dL/U."""
    adjustments = TREND_ADJUSTMENTS[trend]
    if correction_factor < 25:
        return adjustments[0]
    if correction_factor < 50:
        return adjustments[1]
    if correction_factor <= 75:
        return adjustments[2]
    return adjustments[3]


def calculate_suggested_dose(
    *,
    glucose: float,
    carbohydrates: float,
    target_glucose: float,
    correction_factor: float,
    carbohydrate_ratio: float,
    trend: GlucoseTrend,
) -> DoseCalculation:
    """Calcula MAX(0, Round(((glicemia - meta)/FC) + (CHO/RIC) + tendência))."""
    if glucose <= 0 or carbohydrates < 0:
        raise ValueError("Glicemia e carboidratos inválidos.")
    if correction_factor <= 0 or carbohydrate_ratio <= 0:
        raise ValueError("Fatores de insulina devem ser maiores que zero.")
    correction = (glucose - target_glucose) / correction_factor
    coverage = carbohydrates / carbohydrate_ratio
    adjustment = trend_adjustment(trend, correction_factor)
    total = correction + coverage + adjustment
    return DoseCalculation(
        correction, coverage, adjustment, total, max(0, round_half_away_from_zero(total))
    )
